### ISSUE TITLE
FAILED: akka.http.javadsl.server.HttpAppSpec

### ISSUE THREAD

**Issue by jrudolph** [akka/akka-http#1061]
https://jenkins.akka.io:8498/job/pr-validator-akka-http/995/testReport/junit/akka.http.javadsl.server/HttpAppSpec/HttpApp_Java_should_provide_binding_if_available/

"akka.http.javadsl.server.HttpAppSpec.HttpApp Java should provide binding if available"

`The future returned an exception of type: akka.stream.StreamTcpException, with message: Tcp command [Connect(localhost:36181,None,List(),Some(10 seconds),true)] failed.`

**Comment by jrudolph** [comment]
@jlprat do you have an idea what this could be?

**Comment by jlprat** [comment]
I'll take a look, but it smells like race condition trying to bind to whatever port it was meant to bind

**Comment by jlprat** [comment]
So, it fails on the last line of these:
```
      minimal.bindingPromise.get(5, TimeUnit.SECONDS)
      minimal.binding().localAddress.getPort should ===(port)
      minimal.binding().localAddress.getHostName should ===(host)
      // Requesting the server to shutdown
      callAndVerify(host, port, "shutdown")
```
which is:
```
  private def callAndVerify(host: String, port: Int, path: String) = {
    implicit val mat = ActorMaterializer()
    val request = HttpRequest(uri = s"http://$host:$port/$path")
    val response = Http().singleRequest(request)
    response.futureValue.status should ===(StatusCodes.OK)
  }
```
So the binding completed correctly, but for whatever reason this same server wasn't listening. Something is not adding up, as if the promise is completed, the server should be up and running. (promise is completed when the `Future` of `bindAndHandle` is completed.

Will try to find some time to investigate further

**Comment by jrudolph** [comment]
I added https://github.com/akka/akka/pull/22954 which should give us better error messages in the future.

**Comment by jrudolph** [comment]
I also worked a bit on improving `TestUtils` to use all possible loopback interfaces to make random port conflicts less likely. See https://github.com/jrudolph/akka-http/tree/jr/w/choose-random-loopback-interface

Though, after going through all those "Tcp command Connect failed" tickets I'm actually not convinced that port conflicts are actually the reason. Also it doesn't seem to match up with what you reported about how it works here, @jlprat. So, maybe let's wait until the error messages have been improved to see what the cause actually is.

**Comment by jlprat** [comment]
Let's wait till next failure and hopefully we get some more info :+1:

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/pr-validator-akka-http/1077/testReport/junit/akka.http.javadsl.server/HttpAppSpec/HttpApp_Java_should_provide_binding_if_available/

**Comment by jlprat** [comment]
Brief analysis of what happened, it could not open a server on `localhost` port `36181`. Host and port are retrieved using `TestUtils.temporaryServerHostnameAndPort()`

**Comment by jrudolph** [comment]
Where do you see that the server could not be bound? It has a line

```
[INFO] [05/10/2017 07:43:37.604] [ForkJoinPool.commonPool-worker-3] [akka.actor.ActorSystemImpl(MinimalHttpApp)] Server online at http://localhost:36181/
```

**Comment by jlprat** [comment]
I can't see it directly, but the code that fails is https://github.com/akka/akka-http/blob/master/akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala#L174 called from https://github.com/akka/akka-http/blob/master/akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala#L110
So there is no server listening on that port. However, now that I'm taking a deeper look, this test only does one post that actually shuts down the server. Maybe we have some race on the test when the server is asked to shut down... I'll dig deeper

**Comment by ktoso** [comment]
Same thing: https://jenkins.akka.io:8498/job/pr-validator-akka-http/1163/testReport/junit/akka.http.javadsl.server/HttpAppSpec/HttpApp_Java_should_let_get_notified_when_after_binding_is_successful/

**Comment by 2m** [comment]
Fixed in #1207
